apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/RetriableTransaction.cs

```diff
@@ -0,0 +1,135 @@
+            _options = options ?? RetriableTransactionOptions.CreateDefault();
+        }
+
+        public async Task<TResult> RunAsync<TResult>(Func<SpannerTransaction, Task<TResult>> asyncWork, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+
+            // Session will be initialized and subsequently modified by CommitAttempt.
+            PooledSession session = null;
+            try
+            {
+                return await ExecuteWithRetryAsync(CommitAttempt, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                session?.Dispose();
+            }
+
+            async Task<TResult> CommitAttempt()
+            {
+                return await ExecuteHelper.WithErrorTranslationAndProfiling(
+                    async () =>
+                    {
+                        SpannerTransaction transaction = null;
+
+                        try
+                        {
+                            session = await (session?.WithFreshTransactionOrNewAsync(cancellationToken) ?? _connection.AcquireReadWriteSessionAsync(cancellationToken)).ConfigureAwait(false);
+                            transaction = new SpannerTransaction(_connection, TransactionMode.ReadWrite, session, null);
+
+                            TResult result = await asyncWork(transaction).ConfigureAwait(false);
+                            await transaction.CommitAsync(cancellationToken).ConfigureAwait(false);
+
+                            return result;
+                        }
+                        catch when (transaction != null)
+                        {
+                            try
+                            {
+                                await transaction.RollbackAsync(cancellationToken).ConfigureAwait(false);
+                            }
+                            catch (Exception e)
+                            {
+                                // If the commit actually aborted, calling Rollback will fail.
+                                // We don't want that or any other rollback exception to propogate,
+                                // it will not trigger the retry
+                                _connection.Logger.Warn("A rollback attempt failed on RetriableTransaction.RunAsync.CommitAttempt", e);
+                            }
+
+                            // Throw, the retry helper will know when to retry.
+                            throw;
+                        }
+                        finally
+                        {
+                            if (transaction != null)
+                            {
+                                // Let's make sure that the associated session is not released to the pool
+                                // because we'll be attempting to get a fresh transaction for this same session first.
```

Comment:
A few concerns here:

- Don't we want to release the session if the transaction has succeeded? We'll still be executing this code.
- What about if the transaction failed, but we're not going to retry either due to deadlines or it not being an abort?

Reply:
Ah, my guess is that both of these are handled by the final `session?.Dispose()` call - please add a comment to avoid silly people like me getting confused again :)

Reply:
Done :)